pkg/integration/helpers.go

```diff
@@ -0,0 +1,343 @@
+	if err != nil {
+		tb.Errorf("failed to create server: %w", err)
+	}
+
+	// Stop the server on cleanup
+	stopCtx, stop := context.WithCancel(ctx)
+	tb.Cleanup(stop)
+
+	go func() {
+		if err := srv.ServeHTTPHandler(stopCtx, handlers.CombinedLoggingHandler(os.Stdout, adminRouter)); err != nil {
+			tb.Error(err)
+		}
+	}()
+	return srv
+}
+
+func (s *Suite) newAPIServer(ctx context.Context, tb testing.TB) *server.Server {
+	// Create the renderer
+	h, err := render.New(ctx, "", s.cfg.APISrvConfig.DevMode)
+	if err != nil {
+		tb.Errorf("failed to create the renderer %v", err)
+	}
+
+	// Setup cacher
+	cacher, err := cache.CacherFor(ctx, &s.cfg.APISrvConfig.Cache, cache.MultiKeyFunc(
+		cache.HMACKeyFunc(sha1.New, s.cfg.APISrvConfig.Cache.HMACKey),
+		cache.PrefixKeyFunc("apiserver:cache:"),
+	))
+	if err != nil {
+		tb.Fatalf("failed to create cacher: %v", err)
+	}
+	tb.Cleanup(func() {
+		if err := cacher.Close(); err != nil {
+			tb.Errorf("got err when cleanup: %v", err)
+		}
+	})
+
+	// Setup signers
+	tokenSigner, err := keys.KeyManagerFor(ctx, &s.cfg.APISrvConfig.TokenSigning.Keys)
+	if err != nil {
+		tb.Fatalf("failed to create token key manager: %v", err)
+	}
+	certificateSigner, err := keys.KeyManagerFor(ctx, &s.cfg.APISrvConfig.CertificateSigning.Keys)
+	if err != nil {
+		tb.Fatalf("failed to create certificate key manager: %v", err)
+	}
+
+	apiRouter := mux.NewRouter()
+	// Install common security headers
+	apiRouter.Use(middleware.SecureHeaders(ctx, s.cfg.APISrvConfig.DevMode, "json"))
+
+	apiRouter.Handle("/health", controller.HandleHealthz(ctx, &s.cfg.APISrvConfig.Database, h)).Methods("GET")
+
+	{
+		sub := apiRouter.PathPrefix("/api").Subrouter()
+
+		// Setup API auth
+		requireAPIKey := middleware.RequireAPIKey(ctx, cacher, s.db, h, []database.APIUserType{
+			database.APIUserTypeDevice,
+		})
+		// Install the APIKey Auth Middleware
+		sub.Use(requireAPIKey)
+
+		verifyChaff := chaff.New()
+		defer verifyChaff.Close()
+		verifyapiController, err := verifyapi.New(ctx, &s.cfg.APISrvConfig, s.db, h, tokenSigner)
+		if err != nil {
+			tb.Errorf("failed to create verify api controller: %w", err)
+		}
+		sub.Handle("/verify", verifyapiController.HandleVerify()).Methods("POST")
+
+		certChaff := chaff.New()
+		defer certChaff.Close()
+		certapiController, err := certapi.New(ctx, &s.cfg.APISrvConfig, s.db, cacher, certificateSigner, h)
+		if err != nil {
+			tb.Errorf("failed to create certapi controller: %w", err)
+		}
+		sub.Handle("/certificate", certapiController.HandleCertificate()).Methods("POST")
+	}
+
+	srv, err := server.New(s.cfg.APISrvConfig.Port)
+	if err != nil {
+		tb.Errorf("failed to create server: %w", err)
+	}
+
+	// Stop the server on cleanup
+	stopCtx, stop := context.WithCancel(ctx)
+	tb.Cleanup(stop)
+
+	go func() {
+		if err := srv.ServeHTTPHandler(stopCtx, handlers.CombinedLoggingHandler(os.Stdout, apiRouter)); err != nil {
+			tb.Error(err)
+		}
+	}()
+	return srv
```

Comment:
This should be closed when done

Reply:
Same here, the server should be closed by the cancelled context.